### PR TITLE
teensy-cmake-macros: init at unstable-2023-04-15

### PR DESCRIPTION
**Pull request by michaeldonovan**
## Description of changes

Init package for [newdigate/teensy-cmake-macros](https://github.com/newdigate/teensy-cmake-macros/), a set of CMake macros for building [Teensy](https://www.pjrc.com/teensy/) apps/libraries with gcc-arm-none-eabi.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -11048,6 +11048,12 @@
     github = "michaelCTS";
     githubId = 132582212;
   };
+  michaeldonovan = {
```
Comment (AndersonTorres) [inline maintainers/maintainer-list.nix:11051]: Split this in a separate commit.
No need to open a PR.

File: pkgs/development/libraries/teensy-cmake-macros/default.nix [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{ stdenv, lib, fetchFromGitHub, cmake, pkg-config }:
+
+stdenv.mkDerivation {
+  pname = "teensy-cmake-macros";
+  version = "0.1.0";
+
+  src = fetchFromGitHub {
+    owner = "newdigate";
+    repo = "teensy-cmake-macros";
+    rev = "dc401ed23e6e13a9db3cd2a65f611a4738df3b0e";
```
Comment (AndersonTorres) [inline pkgs/development/libraries/teensy-cmake-macros/default.nix:10]: Why aren't you using `version` here?

File: pkgs/development/libraries/teensy-cmake-macros/default.nix [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    hash = "sha256-E+BOlsCJtOScr3B5GSv1WM6rFv6cFYvm/iJ893fsmXM=";
+  };
+
+  propagatedBuildInputs = [ cmake pkg-config ];
+
+  setupHook = ./setup-hook.sh;
```
Comment (AndersonTorres) [inline pkgs/development/libraries/teensy-cmake-macros/default.nix:16]: I suggest you to put this in a passthru.
Look at our code packaging Zig 0.11 as an example.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/teensy-cmake-macros/default.nix (hunk 2 above)
Comment (michaeldonovan) [reply pkgs/development/libraries/teensy-cmake-macros/default.nix:10]: The upstream repo isn't tagged unfortunately. Is there a better way to represent that in the package version?

File: pkgs/development/libraries/teensy-cmake-macros/default.nix (hunk 3 above)
Comment (michaeldonovan) [reply pkgs/development/libraries/teensy-cmake-macros/default.nix:16]: Thanks, I'll take a look.

File: pkgs/development/libraries/teensy-cmake-macros/default.nix (hunk 2 above)
Comment (AndersonTorres) [reply pkgs/development/libraries/teensy-cmake-macros/default.nix:10]: Unfortunately, no one that I know.

File: pkgs/development/embedded/teensy-cmake-macros/default.nix [hunk 4]
```diff
@@ -0,0 +1,27 @@
+{ stdenv, lib, fetchFromGitHub, cmake, pkg-config }:
+
+stdenv.mkDerivation {
+  pname = "teensy-cmake-macros";
+
+  version = "unstable-2023-04-15";
```
Comment (AndersonTorres) [inline pkgs/development/embedded/teensy-cmake-macros/default.nix:6]:
```suggestion
  version = "unstable-2023-04-15";

```

File: pkgs/development/embedded/teensy-cmake-macros/default.nix [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{ stdenv, lib, fetchFromGitHub, cmake, pkg-config }:
+
+stdenv.mkDerivation {
```
Comment (AndersonTorres) [inline pkgs/development/embedded/teensy-cmake-macros/default.nix:3]:
```suggestion
stdenv.mkDerivation (finalAttrs: {
```

File: pkgs/development/embedded/teensy-cmake-macros/default.nix [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    repo = "teensy-cmake-macros";
+    rev = "dc401ed23e6e13a9db3cd2a65f611a4738df3b0e";
+    hash = "sha256-E+BOlsCJtOScr3B5GSv1WM6rFv6cFYvm/iJ893fsmXM=";
+  };
+
+  propagatedBuildInputs = [ cmake pkg-config ];
+
+  passthru = {
+    hook = ./setup-hook.sh;
+  };
```
Comment (AndersonTorres) [inline pkgs/development/embedded/teensy-cmake-macros/default.nix:18]:
```suggestion
  passthru = {
    hook = callPackage ./hook.nix {
      teensy-cmake-macros = finalAttrs.finalPackage;
    };
  };
```

Comment (AndersonTorres) [reply pkgs/development/embedded/teensy-cmake-macros/default.nix:18]: And write something like

```nix
{ lib
, makeSetupHook
, teensy-cmake-macros
}:

makeSetupHook {
  name = "teensy-cmake-macros-hook";

  propagatedBuildInputs = [ teensy-cmake-macros ];

  passthru = { inherit teensy-cmake-macros; };

  meta = {
    description = "A setup hook for teensy-cmake-macros";
    inherit (teensy-cmake-macros.meta) maintainers platforms broken;
  };
} ./setup-hook.sh
```

File: pkgs/development/embedded/teensy-cmake-macros/setup-hook.sh [hunk 7]
```diff
@@ -0,0 +1,4 @@
+_teensyCMakeMacrosEnvHook() {
```
Comment (AndersonTorres) [inline pkgs/development/embedded/teensy-cmake-macros/setup-hook.sh:1]:
```suggestion
teensyCMakeMacrosEnvHook() {
```

File: pkgs/development/embedded/teensy-cmake-macros/setup-hook.sh [hunk 8]
```diff
@@ -0,0 +1,4 @@
...
+  cmakeFlagsArray+=(-DCMAKE_MODULE_PATH=@out@/lib/cmake)
+}
+addEnvHooks "$targetOffset" _teensyCMakeMacrosEnvHook
```
Comment (AndersonTorres) [inline pkgs/development/embedded/teensy-cmake-macros/setup-hook.sh:4]:
```suggestion

addEnvHooks "$targetOffset" teensyCMakeMacrosEnvHook
```

File: pkgs/top-level/all-packages.nix [hunk 9]
```diff
@@ -20103,6 +20103,8 @@ with pkgs;
 
   tcptrack = callPackage ../development/tools/misc/tcptrack { };
 
+  teensy-cmake-macros = callPackage ../development/embedded/teensy-cmake-macros {};
```
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:20106]:
```suggestion
  teensy-cmake-macros = callPackage ../development/embedded/teensy-cmake-macros { };
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/embedded/teensy-cmake-macros/default.nix [hunk 10]
```diff
@@ -0,0 +1,34 @@
+{ stdenv
+, lib
+, fetchFromGitHub
+, cmake
+, pkg-config
+, callPackage }:
```
Comment (AndersonTorres) [inline pkgs/development/embedded/teensy-cmake-macros/default.nix:6]:
```suggestion
, callPackage
}:
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/embedded/teensy-cmake-macros/hook.nix [hunk 11]
```diff
@@ -0,0 +1,18 @@
+{ lib
+, makeSetupHook
+, teensy-cmake-macros
+}:
+
+makeSetupHook
+{
```
Comment (AndersonTorres) [inline pkgs/development/embedded/teensy-cmake-macros/hook.nix:7]:
```suggestion
makeSetupHook {
```

Consistency.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Review by AndersonTorres — APPROVED** [review]
